### PR TITLE
feat(jiraserver) Add OAuth connection process

### PR DESCRIPTION
**Pull request by markstory**
Build out the OAuth1 connection flow required for Jira-Server. Jira uses a RSA key signed OAuth1 system. I've added the required oauth libraries to our requirements, however this library was already installed in getsentry.

I've chosen not to implement `sentry.identity.provider` interface as there is no requirement for Jira to be used as a single signon provider and the additional IdentityProviderPipeline could would go unused.

Refs APP-805

### CODE REVIEW

File: src/sentry/integrations/jira_server/integration.py [hunk 1]
```diff
@@ -132,8 +138,70 @@ def dispatch(self, request, pipeline):
...
+
+
+class OAuthCallbackView(PipelineView):
+    """
+    Complete the OAuth dance by exchanging our request token
+    into an access token.
+    """
+    @csrf_exempt
+    def dispatch(self, request, pipeline):
+        config = pipeline.state.data['installation_data']
```
Comment (lauryndbrown) [inline src/sentry/integrations/jira_server/integration.py]: So you can do that, but you can also do:
```python
pipeline.fetch_state('installation_data')
```
pretty sure those are equivalent statements

File: src/sentry/integrations/jira_server/integration.py [hunk 2]
```diff
@@ -132,8 +138,70 @@ def dispatch(self, request, pipeline):
...
 
 
+class OAuthLoginView(PipelineView):
+    """
+    Start the OAuth dance by creating a request token
+    and redirecting the user to approve it.
+    """
+    @csrf_exempt
+    def dispatch(self, request, pipeline):
+        if 'oauth_token' in request.GET:
```
Comment (lauryndbrown) [inline src/sentry/integrations/jira_server/integration.py]: Does the user have an accept/reject button to press? What happens here if the user doesn't authorize us?

File: src/sentry/integrations/jira_server/integration.py [hunk 3]
```diff
@@ -156,43 +224,35 @@ class JiraServerIntegrationProvider(IntegrationProvider):
...
+            'external_id': '%s:%s' % (hostname, install['consumer_key']),
+            'metadata': {
+                'base_url': install['url'],
+                'verify_ssl': install['verify_ssl'],
+            },
             'user_identity': {
                 'type': 'jira_server',
-                'external_id': '%s:%s' % (state['base_url'], user['id'])
+                'external_id': '%s:%s' % (hostname, install['consumer_key']),
+                'scopes': (),
```
Comment (lauryndbrown) [inline src/sentry/integrations/jira_server/integration.py]: I was curious, are there no oauth scopes in jira server?

**Review by lauryndbrown — APPROVED** [review]
Just additional thoughts here.

File: src/sentry/integrations/jira_server/integration.py (hunk 2 above)
Comment (markstory) [reply src/sentry/integrations/jira_server/integration.py]: The user does have a deny/allow button inside Jira. If they click deny, they are sent back to us with an `oauth_token` in the GET params. When we try to fetch the access token we get a 401 error.

Right now we display 'Could not fetch an access token from Jira' in the pipeline error page.

File: src/sentry/integrations/jira_server/integration.py (hunk 3 above)
Comment (markstory) [reply src/sentry/integrations/jira_server/integration.py]: No. Jira's Oauth1.0a implementation doesn't have scopes.

File: src/sentry/integrations/jira_server/integration.py (hunk 1 above)
Comment (markstory) [reply src/sentry/integrations/jira_server/integration.py]: Which do you think we should be using? I don't have a preference.

Comment (EvanPurkhiser) [reply src/sentry/integrations/jira_server/integration.py]: Prefer `fetch_state` call

File: src/sentry/integrations/jira_server/integration.py (hunk 2 above)
Comment (lauryndbrown) [reply src/sentry/integrations/jira_server/integration.py]: That's fine. The user took a direct action so there shouldn't be confusion there.

**Pull request merged**